File: typescript/src/core/context/pipeline/Step.ts
import type { UContextType } from "../type";

export type TStepExecutor<TContext, TResult> = (
  ctx: TContext,
) => Promise<TResult>;

/**
 * Lớp định nghĩa bước xử lý trong một pipeline
 *
 * @template TResult
 */
export class Step<TContext = unknown, TResult = unknown> {
  public ctxType: UContextType;

  private _executor?: TStepExecutor<TContext, TResult>;

  constructor(
    ctxType: UContextType = "runtime",
    executor?: TStepExecutor<TContext, TResult>,
  ) {
    this.ctxType = ctxType;

    if (executor) this._executor = executor;
  }

  /**
   * Gán executor cho Step
   *
   * @param executor - là một hàm
   */
  setExecutor(executor: TStepExecutor<TContext, TResult>) {
    this._executor = executor;
  }

  /**
   * Thực thi executor của một Step
   *
   * @param ctx - ngữ cảnh thực thi, có thể là runtime (TRuntimeContext)
   * hoặc trong core (TInternalContext)
   */
  execute(ctx: TContext) {
    if (!this._executor) throw new Error("Executor of Step is not set");
    return this._executor(ctx);
  }
}

File: python/src/core/context/pipeline/Step.py
from typing import Generic, Optional, Awaitable, Union

from .type import TContext, TResult, TStepExecutor


class Step(Generic[TContext, TResult]):
    """
    Lớp định nghĩa bước xử lý trong một pipeline
    """

    def __init__(self, executor: Optional[TStepExecutor[TContext, TResult]] = None):
        self._executor: Optional[TStepExecutor[TContext, TResult]] = None
        if executor:
            self._executor = executor

    def set_executor(self, executor: TStepExecutor[TContext, TResult]) -> None:
        """
        Gán executor cho Step

        Args:
            executor: là một hàm
        """
        self._executor = executor

    def execute(self, ctx: TContext) -> Union[TResult, Awaitable[TResult]]:
        """
        Thực thi executor của một Step

        Args:
            ctx: ngữ cảnh thực thi, có thể là runtime (RuntimeContext)
                 hoặc trong core (InternalContext)
        """
        if not self._executor:
            raise ValueError("Executor of Step is not set")

        return self._executor(ctx)
